src/main/scala/analysis/MemoryRegionAnalysis.scala

```diff
@@ -195,58 +215,109 @@ trait MemoryRegionAnalysis(val program: Program,
 //          }
           if (directCall.target.name == "malloc") {
             evaluateExpression(mallocVariable, constantProp(n)) match {
-              case Some(b: BitVecLiteral) => regionLattice.lub(s, Set(HeapRegion(nextMallocCount(), b, IRWalk.procedure(n))))
-              case None => s
+              case Some(b: BitVecLiteral) =>
+                val negB = if isNegative(b) then b.value - BigInt(2).pow(b.size) else b.value
+                val (name, start) = nextMallocCount(negB)
+                val newHeapRegion = HeapRegion(name, start, negB, IRWalk.procedure(n))
+                addReturnHeap(directCall, newHeapRegion)
+                s
+              case None =>
+                // Assume heap region size is at least 1 TODO: must approximate size of heap
+                val negB = 1
+                val (name, start) = nextMallocCount(negB)
+                val newHeapRegion = HeapRegion(name, start, negB, IRWalk.procedure(n))
+                addReturnHeap(directCall, newHeapRegion)
+                s
```

Comment:
You don't need to approximate the size of a heap allocation if you don't know it. The size of each heap allocation is not important except for detecting out-of-bounds accesses (which is not the most important). Each heap allocation needs to be its own separate address space, with the R0 returned from malloc being like the stack pointer for that heap.